File: src/test/java/dev/jiaming/ai_interview/coach/CoachResponseMapperTests.java
package dev.jiaming.ai_interview.coach;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.util.List;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import dev.jiaming.ai_interview.gemini.GeminiException;

class CoachResponseMapperTests {

	private final CoachResponseMapper mapper = new CoachResponseMapper(new ObjectMapper());

	@Test
	void rejectsQuestionResponsesWithoutUsableQuestions() {
		InterviewQuestionsResponse response = new InterviewQuestionsResponse(
			List.of(new InterviewQuestionResponse("", "", "", "  ", List.of(), List.of())),
			"gemini"
		);

		assertThatThrownBy(() -> mapper.normalizeQuestions(response, List.of("resume:experience:0")))
			.isInstanceOf(GeminiException.class)
			.hasMessage("Gemini returned no usable interview questions");
	}

	@Test
	void normalizesAUsableQuestion() {
		InterviewQuestionsResponse response = new InterviewQuestionsResponse(
			List.of(new InterviewQuestionResponse(
				null,
				"System Design",
				"deep",
				"How would you make this workflow idempotent?",
				List.of("Unique operation keys"),
				List.of()
			)),
			"Gemini"
		);

		InterviewQuestionsResponse normalized = mapper.normalizeQuestions(response, List.of("resume:experience:0"));

		assertThat(normalized.questions()).hasSize(1);
		assertThat(normalized.questions().getFirst().difficulty()).isEqualTo("Deep Dive");
		assertThat(normalized.questions().getFirst().sourceContextIds()).containsExactly("resume:experience:0");
	}

	@Test
	void rejectsSourceContextIdsThatWereNotRetrieved() {
		InterviewQuestionsResponse response = new InterviewQuestionsResponse(
			List.of(new InterviewQuestionResponse(
				"question-id",
				"Projects",
				"Core",
				"How did you design this project?",
				List.of("Architecture decisions"),
				List.of("resume:4", "invented:source:9")
			)),
			"gemini"
		);

		InterviewQuestionsResponse normalized = mapper.normalizeQuestions(
			response,
			List.of("resume:projects:4", "resume:skills:5")
		);

		assertThat(normalized.questions().getFirst().sourceContextIds())
			.containsExactly("resume:projects:4", "resume:skills:5");
	}
}

File: src/test/kotlin/dev/jiaming/ai_interview/coach/CoachResponseMapperTests.kt
package dev.jiaming.ai_interview.coach

import com.fasterxml.jackson.databind.ObjectMapper
import dev.jiaming.ai_interview.gemini.GeminiException
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test

class CoachResponseMapperTests {
	private val mapper = CoachResponseMapper(ObjectMapper())

	@Test
	fun rejectsQuestionResponsesWithoutUsableQuestions() {
		val response = InterviewQuestionsResponse(listOf(InterviewQuestionResponse("", "", "", "  ", emptyList(), emptyList())), "gemini")
		assertThatThrownBy { mapper.normalizeQuestions(response, listOf("resume:experience:0")) }
			.isInstanceOf(GeminiException::class.java)
			.hasMessage("Gemini returned no usable interview questions")
	}

	@Test
	fun normalizesAUsableQuestion() {
		val response = InterviewQuestionsResponse(listOf(InterviewQuestionResponse(null, "System Design", "deep", "How would you make this workflow idempotent?", listOf("Unique operation keys"), emptyList())), "Gemini")
		val normalized = mapper.normalizeQuestions(response, listOf("resume:experience:0"))
		assertThat(normalized.questions()).hasSize(1)
		assertThat(normalized.questions().first().difficulty()).isEqualTo("Deep Dive")
		assertThat(normalized.questions().first().sourceContextIds()).containsExactly("resume:experience:0")
	}

	@Test
	fun rejectsSourceContextIdsThatWereNotRetrieved() {
		val response = InterviewQuestionsResponse(listOf(InterviewQuestionResponse("question-id", "Projects", "Core", "How did you design this project?", listOf("Architecture decisions"), listOf("resume:4", "invented:source:9"))), "gemini")
		val normalized = mapper.normalizeQuestions(response, listOf("resume:projects:4", "resume:skills:5"))
		assertThat(normalized.questions().first().sourceContextIds()).containsExactly("resume:projects:4", "resume:skills:5")
	}
}
